README.md
# Velero-Watchdog-Pro

> [!WARNING]  
**Attention Users:** This project is in active development, and certain tools or features might still be under construction. We kindly urge you to exercise caution while utilizing the tools within this environment. While every effort is being made to ensure the stability and reliability of the project, there could be unexpected behaviors or limited functionalities in some areas.
We highly recommend thoroughly testing the project in non-production or sandbox environments before implementing it in critical or production systems. Your feedback is invaluable to us; if you encounter any issues or have suggestions for improvement, please feel free to [report them](https://github.com/seriohub/velero-watchdog-pro/issues). Your input helps us enhance the project's performance and user experience.
Thank you for your understanding and cooperation.

## Description

This Python project is designed to monitor the status of Velero backups in Kubernetes environments and alert when something is not working.
It provides efficient ways to check the backups status.
The user chooses the synchronization cycles and the elements to monitor.

## Features

### 1. Backup Status Monitoring

The project monitors the backup status of Kubernetes clusters.

### 2. Schedule Change Monitoring

Monitor and alert if the schedule changes.

### 3. Channels notifications

Receive the alerts and the solved messages via notifications channels, allowing immediate action.

Available plugin:
- Telegram
- email


## Requirements

- Python 3.x
- kubectl cli (if [Run in kubernetes](#run-in-kubernetes))
- Telegram API credentials (if telegram notification is enabled)
- SMTP and user account (if email notification is enabled)

## Configuration

| FIELD                       | TYPE   | DEFAULT | DESCRIPTION                                                                                                                                              |
|-----------------------------|--------|---------|----------------------------------------------------------------------------------------------------------------------------------------------------------|
| `DEBUG`                     | Bool   | False   | View debugging information.                                                                                                                              |
| `LOG_SAVE`                  | Bool   | False   | Save log to files                                                                                                                                        |
| `PROCESS_LOAD_KUBE_CONFIG`* | Bool   | True    | Set False if it runs on k8s.                                                                                                                             |
| `PROCESS_KUBE_CONFIG`       | String |         | Path to the kube config file. This is mandatory when the script runs outside the Kubernetes cluster, either in a docker container or as a native script. |
| `PROCESS_CLUSTER_NAME` * ** | String |         | Force the cluster name and it appears in the telegram message                                                                                            |
| `PROCESS_CYCLE_SEC`         | Int    | 120     | Cycle time (seconds)                                                                                                                                     |
| `TELEGRAM_ENABLE`    *      | Bool   | True    | Enable telegram notification                                                                                                                             |
| `TELEGRAM_API_TOKEN` *      | String |         | Token for access to Telegram bot via Http API                                                                                                            |
| `TELEGRAM_CHAT_ID`   *      | String |         | Telegram chat id where send the notifications                                                                                                            |
| `EMAIL_ENABLE`       *      | Bool   | False   | Enable email notification                                                                                                                                |
| `EMAIL_SMTP_SERVER`  *      | String |         | SMTP server                                                                                                                                              |
| `EMAIL_SMTP_PORT`    *      | int    | 587     | SMTP port                                                                                                                                                |
| `EMAIL_ACCOUNT`      *      | String |         | user name account                                                                                                                                        |
| `EMAIL_PASSWORD`     *      | String |         | password account                                                                                                                                         |
| `EMAIL_RECIPIENTS`   *      | Bool   |         | Email recipients                                                                                                                                         |
| `BACKUP_ENABLE`             | Bool   | True    | Enable watcher for backups without schedule or last backup for each schedule                                                                             |
| `EXPIRES_DAYS_WARNING`      | int    | 29      | Number of days to backup expiration below which to display a warning about the backup                                                                    |
| `SCHEDULE_ENABLE`           | Bool   | True    | Enable watcher for schedule                                                                                                                              |
| `K8S_INCLUSTER_MODE` **     | Bool   | False   | Enable in cluster mode                                                                                                                                   |
| `IGNORE_NM_1`               | String |         | regex to ignore a namespace or a group of namespaces                                                                                                     |
| `IGNORE_NM_2`               | String |         | regex to ignore a namespace or a group of namespaces                                                                                                     |
| `IGNORE_NM_3`               | String |         | regex to ignore a namespace or a group of namespaces                                                                                                     |

*Mandatory parameters<br>
** Mandatory if it is deployed on cluster

<br>If you set "TELEGRAM_ENABLE"=False, the application prints out the message only on the stdout

## Installation

Clone the repository:

  ``` bash
    git clone https://github.com/seriohub/velero-watchdog-pro.git
    cd velero-watchdog-pro
  ```

### Run native

1. Navigate to the [src](src) folder

2. Dependencies installation:

    ``` bash
    pip install -r requirements.txt
    ```

3. Configuration

    Create and edit .env file under src folder, you can start from [.env.template](src/.env.template) under [src](src) folder
    Setup mandatory parameters in the src/.env file if runs it in the native mode

4. Usage

    Run the main script:

    ``` bash
    python3 main.py
    ```

### Run in Docker

1. Configuration
   1. Navigate to the [docker](docker) folder
   2. Setup mandatory parameters in the docker-compose.yaml file ([docker-compose.yaml](docker/docker-compose.yaml))

      **Note:** Instead of editing the docker-compose.yaml file, you can create and edit .env file (you can start from [.env.template](docker/.env.template)) file under [docker](docker) folder and use the docker-compose.yaml default  values.
2. Docker image :
   ##### Build docker image from scratch
      1. Navigate to the root folder
      2. Build the docker image
      ``` bash
      docker build --target velero-watchdog-pro -t velero-watchdog-pro:0.1.0 -t velero-watchdog-pro:latest -f ./docker/Dockerfile .
      ```
   ##### Use image published on DockerHub
      1. Pull the image 
      ``` bash
      docker pull dserio83/velero-watchdog-pro
      ```
3. Create docker volume:
   1. Create a volume for store logs
      ``` bash
      docker volume create velero_watchdog_vol
      ```

4. Create the stack and run it

    ``` bash
     docker-compose  -f ./docker/docker-compose.yaml -p velero-watchdog-stack  up -d
    ```

### Run in Kubernetes

1. Configuration
   1. Navigate to the [k8s](k8s) folder
   2. Create and edit .env file under k8s folder, you can start from [.env.template](k8s/.env.template)
  
   3. Export .env

      ``` bash
      export $(cat .env | xargs)
      ```

      Checks that the variables have been exported:

      ``` bash
      printenv | grep K8SW_* 
      ```
 
2. Setup docker image:

   1. Navigate to the root folder
   2. Build image

        ``` bash
        docker build --target velero-watchdog-pro -t ${K8SW_DOCKER_REGISTRY}/${K8SW_DOCKER_IMAGE}:0.1.0 -t ${K8SW_DOCKER_REGISTRY}/${K8SW_DOCKER_IMAGE}:latest -f ./docker/Dockerfile .
        ```

   3. Push image

        ``` bash
        docker push ${K8SW_DOCKER_REGISTRY}/${K8SW_DOCKER_IMAGE} --all-tags
        ```

      >[!INFO]  
      Alternative you can use skip the *Build image* and *Push image* steps and use an deployed image published on DockerHub.<br>
      Edit the .env file:
      **K8SW_DOCKER_REGISTRY=dserio83** <br>
      More info: https://hub.docker.com/r/dserio83/velero-watchdog-pro

3. Kubernetes create objects

   1. Navigate to the [k8s](k8s) folder

   2. Create namespace:

        ``` bash
        cat 10_create_ns.yaml | envsubst | kubectl apply -f -
        ```

   3. Create the PVC:

       ``` bash
       cat 20_pvc.yaml | envsubst | kubectl apply -f -
       ```

   4. Create the ConfigMap:

       ``` bash
       cat 30_cm.yaml | envsubst | kubectl apply -f -
       ```
  
   5. Create the RBAC:

       ``` bash
        cat 40_rbac.yaml | envsubst | kubectl apply -f -
       ```
  
   6. Create the Deployment:

       ``` bash
        cat 50_deployment.yaml | envsubst | kubectl apply -f -
       ```

## Test Environment

The project is developed, tested and put into production on several clusters with the following configuration

1. Kubernetes v1.28.2

## How to Contribute

1. Fork the project
2. Create your feature branch

    ``` bash
    git checkout -b feature/new-feature
    ```

3. Commit your changes

    ``` bash
   git commit -m 'Add new feature'
   ```

4. Push to the branch

    ``` bash
   git push origin feature/new-feature
   ```

5. Create a new pull request

## License

This project is licensed under the [MIT License](LICENSE).

---

Feel free to modify this template according to your project's specific requirements.

In case you need more functionality, create a PR. If you find a bug, open a ticket.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## [0.1.3] - 2023-11-26
**Implemented enhancements:**
- Add ignored namespace


## [0.1.2] - 2023-11-19
**Implemented enhancements:**
- Run a pre-batch data for avoiding unuseful messages/warnings. <br>When the backup expiration data is grater than the thredshold value, force they to thredshold value

## [0.1.1] - 2023-11-08

**Implemented enhancements:**
- Added cluster name in the report details

### Fixed
- The alive message is sent directly to the queue


### First Release

v0.1.0: [https://github.com/seriohub/velero-watchdog-pro/releases/tag/v0.1.0](https://github.com/seriohub/velero-watchdog-pro/releases/tag/v0.1.0)

src/utils/config.py
from dotenv import load_dotenv
import os
from utils.handle_error import handle_exceptions_static_method, handle_exceptions_method


# class syntax
class ConfigProgram:
    def __init__(self, path_env=None, debug_on=True):
        self.debug_on = debug_on
        res = load_dotenv(dotenv_path=path_env)
        print(f"INFO    [Env] Load env={res}")

    @handle_exceptions_static_method
    def load_key(self, key, default, print_out: bool = True, mask_value: bool = False):
        value = os.getenv(key)
        if value is None or \
                len(value) == 0:
            value = default

        if print_out:
            if mask_value and len(value) > 2:
                index = int(len(value) / 2)
                partial = '*' * index
                if self.debug_on:
                    print(f"INFO    [Env] load_key.key={key} value={value[:index]}{partial}")
            else:
                if self.debug_on:
                    print(f"INFO    [Env] load_key.key={key} value={value}")

        return value

    @handle_exceptions_method
    def logger_key(self):
        return self.load_key('LOG_KEY', 'k8s-wdt')

    @handle_exceptions_method
    def logger_msg_format(self):
        default = '%(asctime)s :: [%(levelname)s] :: %(message)s'
        return self.load_key('LOG_FORMAT', default)

    @handle_exceptions_method
    def logger_save_to_file(self):
        res = self.load_key('LOG_SAVE', 'False')
        return True if res.upper() == 'TRUE' else False

    @handle_exceptions_method
    def logger_folder(self):
        return self.load_key('LOG_DEST_FOLDER',
                             './logs')

    @handle_exceptions_method
    def logger_filename(self):
        return self.load_key('LOG_FILENAME',
                             'k8s.log')

    @handle_exceptions_method
    def logger_max_filesize(self):
        return int(self.load_key('LOG_MAX_FILE_SIZE',
                                 4000000))

    @handle_exceptions_method
    def logger_his_backups_files(self):
        return int(self.load_key('LOG_FILES_BACKUP',
                                 '5'))

    @handle_exceptions_method
    def logger_level(self):
        return int(self.load_key('LOG_LEVEL',
                                 '20'))

    @handle_exceptions_method
    def process_run_sec(self):
        res = self.load_key('PROCESS_CYCLE_SEC',
                            '120')

        if len(res) == 0:
            res = '120'
        return int(res)

    @handle_exceptions_method
    def internal_debug_enable(self):
        res = self.load_key('DEBUG', 'False')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def telegram_token(self):
        return self.load_key('TELEGRAM_TOKEN', '', mask_value=True)

    @handle_exceptions_method
    def telegram_chat_id(self):
        return self.load_key('TELEGRAM_CHAT_ID', '', mask_value=True)

    @handle_exceptions_method
    def telegram_enable(self):
        res = self.load_key('TELEGRAM_ENABLE', 'False')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def telegram_max_msg_len(self):
        res = self.load_key('TELEGRAM_MAX_MSG_LEN',
                            '3000')

        if len(res) == 0:
            res = '2000'
        return int(res)

    @handle_exceptions_method
    def telegram_rate_limit_minute(self):
        res = self.load_key('TELEGRAM_MAX_MSG_MINUTE',
                            '20')

        if len(res) == 0:
            res = '20'
        return int(res)

    @handle_exceptions_method
    def notification_alive_message_hours(self):
        res = self.load_key('NOTIFICATION_ALIVE_MSG_HOURS',
                            '24')
        n_hours = int(res)
        if n_hours < 0:
            n_hours = 0
        elif n_hours > 100:
            n_hours = 100

        return n_hours

    @handle_exceptions_method
    def email_enable(self):
        res = self.load_key('EMAIL_ENABLE', 'False')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def email_smtp_server(self):
        return self.load_key('EMAIL_SMTP_SERVER', '')

    @handle_exceptions_method
    def email_smtp_port(self):
        res = self.load_key('EMAIL_SMTP_PORT',
                            '587')
        n_port = int(res)
        return n_port

    @handle_exceptions_method
    def email_recipient(self):
        return self.load_key('EMAIL_RECIPIENTS', '')

    @handle_exceptions_method
    def email_sender_password(self):
        return self.load_key('EMAIL_PASSWORD', '', mask_value=True)

    @handle_exceptions_method
    def email_sender(self):
        return self.load_key('EMAIL_ACCOUNT', '')

    @handle_exceptions_method
    def k8s_load_kube_config_method(self):
        res = self.load_key('PROCESS_LOAD_KUBE_CONFIG', 'True')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def k8s_config_file(self):
        return self.load_key('PROCESS_KUBE_CONFIG', None)

    @handle_exceptions_method
    def k8s_cluster_identification(self):
        return self.load_key('PROCESS_CLUSTER_NAME', None)

    @handle_exceptions_method
    def k8s_incluster_mode(self):
        res = self.load_key('K8S_INCLUSTER_MODE', 'False')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def velero_backup_enable(self):
        res = self.load_key('BACKUP_ENABLE', 'True')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def velero_schedule_enable(self):
        res = self.load_key('SCHEDULE_ENABLE', 'True')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def velero_expired_days_warning(self):
        res = self.load_key('EXPIRES_DAYS_WARNING',
                            '20')

        if len(res) == 0:
            res = '20'
        return int(res)

    @handle_exceptions_method
    def get_regex_patterns_ignore_nm(self):
        regex_list = []

        for i in range(1, 10):
            res = self.load_key(f'IGNORE_NM_{i}', None)
            if res is not None:
                # Append the regex pattern to the list
                # regex_list.append(re.compile(res))
                regex_list.append(res)
            else:
                break
        return regex_list


class ConfigK8sProcess:
    def __init__(self, cl_config: ConfigProgram = None):
        self.k8s_in_cluster_mode = True
        self.k8s_config_file = None

        self.cluster_name = None
        self.cluster_name_key = 'cluster'

        self.backup_enable = True
        self.backup_key = 'backup'

        self.schedule_enable = True
        self.schedule_key = 'schedule'

        self.disp_msg_key_unique = True  # Fixed True
        self.disp_msg_key_start = 'msg_key_start'
        self.disp_msg_key_end = 'msg_key_end'

        # LS 2023.11.23 add ignored namespaces
        self.ignore_namespace = []

        if cl_config is not None:
            self.__init_configuration_app__(cl_config)

    def __print_configuration__(self):
        """
        Print setup class
        """
        print(f"INFO    [Process setup] k8s cluster name= {self.cluster_name}")

        print(f"INFO    [Process setup] k8s in cluster mode={self.k8s_in_cluster_mode}")
        print(f"INFO    [Process setup] k8s config file={self.k8s_config_file}")
        print(f"INFO    [Process setup] velero backup enable={self.backup_enable}")
        print(f"INFO    [Process setup] velero schedule enable={self.schedule_enable}")
        print(f"INFO    [Process setup] k8s send summary message={self.disp_msg_key_unique}")

        print(f"INFO    [Process setup] k8s ignored namespaces: regex defined {len(self.ignore_namespace)}")

    def __init_configuration_app__(self, cl_config: ConfigProgram):
        """
        Init configuration class reading .env file
        """
        self.backup_enable = cl_config.velero_backup_enable()
        self.schedule_enable = cl_config.velero_schedule_enable()
        self.EXPIRES_DAYS_WARNING = cl_config.velero_expired_days_warning()
        self.cluster_name = cl_config.k8s_cluster_identification()
        self.k8s_in_cluster_mode = cl_config.k8s_incluster_mode()
        self.k8s_config_file = cl_config.k8s_config_file()

        # LS 2023.11.23 add ignored namespace
        self.ignore_namespace = cl_config.get_regex_patterns_ignore_nm()

        self.__print_configuration__()


class ConfigDispatcher:
    def __init__(self, cl_config: ConfigProgram = None):
        self.max_msg_len = 50000
        self.alive_message = 24

        self.telegram_enable = False
        self.telegram_chat_id = '0'
        self.telegram_token = ''
        self.telegram_max_msg_len = 2000
        self.telegram_rate_limit = 20

        self.email_enable = True
        self.email_smtp_server = ''
        self.email_smtp_port = 587
        self.email_sender = ''
        self.email_sender_password = '***'
        self.email_recipient = ''

        if cl_config is not None:
            self.__init_configuration_app__(cl_config)

    def __mask_data__(self, data):
        if len(data) > 2:
            index = int(len(data) / 2)
            partial = '*' * index
            return f"{self.telegram_token[:index]}{partial}"
        else:
            return data

    def __print_configuration__(self):
        """
        Print setup class
        """

        print(f"INFO    [Dispatcher setup] telegram={self.telegram_enable}")
        if self.telegram_enable:
            print(f"INFO    [Dispatcher setup] telegram-chat id={self.telegram_chat_id}")
            print(f"INFO    [Dispatcher setup] telegram-token={self.__mask_data__(self.telegram_token)}")
            print(f"INFO    [Dispatcher setup] telegram-max message length={self.telegram_max_msg_len}")
            print(f"INFO    [Dispatcher setup] telegram-rate limit minute={self.telegram_rate_limit}")
            print(f"INFO    [Dispatcher setup] Notification-alive message every={self.alive_message} hour")

        print(f"INFO    [Dispatcher setup] email={self.email_enable}")
        if self.email_enable:
            print(f"INFO    [Dispatcher setup] email-smtp server={self.email_smtp_server}")
            print(f"INFO    [Dispatcher setup] email-port={self.email_smtp_port}")
            print(f"INFO    [Dispatcher setup] email-sender={self.email_sender}")
            # print(f"INFO    [Dispatcher setup] email-password={self.__mask_data__(self.email_sender_password)}")
            print(f"INFO    [Dispatcher setup] email-password={self.email_sender_password}")
            print(f"INFO    [Dispatcher setup] email-recipient={self.email_recipient}")

    def __init_configuration_app__(self, cl_config: ConfigProgram):
        """
        Init configuration class reading .env file
        """
        # telegram section
        self.telegram_enable = cl_config.telegram_enable()
        self.telegram_chat_id = cl_config.telegram_chat_id()
        self.telegram_token = cl_config.telegram_token()
        self.telegram_max_msg_len = cl_config.telegram_max_msg_len()
        self.telegram_rate_limit = cl_config.telegram_rate_limit_minute()

        self.email_enable = cl_config.email_enable()
        self.email_sender = cl_config.email_sender()
        self.email_sender_password = cl_config.email_sender_password()

        self.email_smtp_port = cl_config.email_smtp_port()
        self.email_smtp_server = cl_config.email_smtp_server()
        self.email_recipient = cl_config.email_recipient()

        self.alive_message = cl_config.notification_alive_message_hours()

        # email section
        self.__print_configuration__()
